### ISSUE TITLE
Feature request: Self-serve DECT preregistration

### ISSUE THREAD

**Issue by kchr** [bornhack/bornhack-website#448]
***TL;DR***

POC team would like to offer pre-registration as a feature on the user profile at Bornhack web site.

# User story

As a hacker visiting camp, I want the Deluxe(tm) Seamless(r) experience of booting my DECT phone without all the hassle and away-from-keyboardness a visit to the POC may induce.

So, until the POC is fully integrated with the bar area I expect our robot overlords to handle all manual labor leading to my phone being ready for novely calls and ordering drinks.

# Background

To enable an automated flow for registration upon arrival, the POC needs:

- The extension number that was requested
- A verification token associated with the extension

The token will be used to register and activate the associated extension number with the device can provide it.

Like the Eventphone registration flow, we will use phone numbers as the verification token.

1. When an extension is reserved by a user, a longer (unique) phone number is generated and associated with the extension number requested.
2. At arrival to the camp, all DECT phones not yet verified are given throwaway numbers that are usable directly.
3. When the user dial the verification number associated with the extension actually requested, it will be activated and registered to the DECT phone that made the call.

The temporary numbers given at arrival have a separate dial plan that can only dial the automated activation numbers, no other handsets. Handsets are sandboxed this way until they are verified by the user (or blessed by the POC).

# Proposal

We would like to request two additional input fields for the user profile model:

- DECT extension (int)
- Publish in phonebook? (bool)

If possible, it would be really nice if there was a way to present the verification number on the user profile page as well.

Could this number be generated per-user in beforehand? It doesn't really have to be related to the actual extension number, since we do not plan to support multiple extensions per user at this point... It could be hidden up until a specific date before camp or so.

It just needs to be unique for each extension and of a reasonable length (10 digits?)

If this is doable we could just pull a CSV file with the extensions and verification numbers a week before the camp and have everything else prepared.

**Comment by tykling** [comment]
see also #334

**Comment by tykling** [comment]
Is it always 4 digit numbers?

Do we want something like a name for the phonebook or do we use the public_credit_name we already have in the profile?

Do we want the number and activation codes and all that to be Camp specific? Like is each new BornHack a new event where we begin from scratch? Or should it be so once I register TYKL (8955) then that is my number for future events until I unregister it?

What about reserved numbers? I assume POC team would need some way to reserve numbers or ranges for services? Or can we use the full range from 0000-9999 (assuming 4 digit numbers)?

**Comment by klondi** [comment]
> Is it always 4 digit numbers?
Yes, smaller digit numbers should still be provided manually (and tied to a normal account) to avoid exhaustion.

> Do we want something like a name for the phonebook or do we use the public_credit_name we already have in the profile?
We want at least a way to explain what the 4 digits mean (i.e. 8955 stands for TYKL). This is an optional field.
 
> Do we want the number and activation codes and all that to be Camp specific? Like is each new BornHack a new event where we begin from scratch? Or should it be so once I register TYKL (8955) then that is my number for future events until I unregister it?
Number activation codes should be camp specific as you may have different hardware from camp to camp (and therefore a different IPEI). About keeping numbers across events I'm neither for nor against. Can you raise this on the Kickoff and see what other volunteers think?

> What about reserved numbers? I assume POC team would need some way to reserve numbers or ranges for services? Or can we use the full range from 0000-9999 (assuming 4 digit numbers)?
We can provide a list of reserved numbers so you can also provide an API and ease wardialing efforts :P

**Comment by tykling** [comment]
> > Is it always 4 digit numbers?
> 
> Yes, smaller digit numbers should still be provided manually (and tied to a normal account) to avoid exhaustion.

what about longer numbers? could I get a 5 or 6 digit number if I should desire one?

**Comment by tykling** [comment]
clarified on irc; longer numbers are okay (but technically only the first 4 are used to route the call)

**Comment by tykling** [comment]
Please discuss in #465

**Comment by tykling** [comment]
What info needs to be included in the data export for the DECT system? Is CSV the best format for it?
